b2/sync/scan_policies.py

```diff
@@ -0,0 +1,157 @@
+#
+######################################################################
+
+import logging
+import re
+import six
+import os.path
+from abc import ABCMeta, abstractmethod
+
+logger = logging.getLogger(__name__)
+
+
+@six.add_metaclass(ABCMeta)
+class ScanExcludePolicy(object):
+    """
+    Policy object that decides which files should be excluded in a scan.
+    Exclude policies never force file. Matched files are dropped irreversibly.
+    """
+
+    @abstractmethod
+    def should_exclude_file(self, file_path):
+        """
+        Given the full path of a file, should it be excluded from the scan?
+
+        :param file_path: The path of the file, relative to the root directory
+                          being scanned.
+        :return: True iff excluded.
+        """
+
+    @abstractmethod
+    def should_exclude_directory(self, dir_path):
+        """
+        Given the full path of a directory, should all of the files in it be
+        excluded from the scan?
+
+        :param dir_path: The path of the directory, relative to the root directory
+                         being scanned.  This is a directory, so the path will
+                         always end in '/'.
+        :return: True iff excluded.
+        """
+
+    def should_exclude(self, path):
+        if path.endswith('/'):
+            return self.should_exclude_directory(path)
+        return self.should_exclude_file(path)
+
+
+@six.add_metaclass(ABCMeta)
+class ScanIncludePolicy(object):
+    """
+    Policy object that decides which files should be included in a scan.
+    Include policies always force matched file. Once forced file is always in scan.
+    """
+
+    @abstractmethod
+    def should_include_file(self, file_path):
+        """
+        Given the full path of a file, should it be included from the scan?
+
+        :param file_path: The path of the file, relative to the root directory
+                          being scanned.
+        :return: True iff included.
+        """
+
+    @abstractmethod
+    def should_include_directory(self, dir_path):
+        """
+        Given the full path of a directory, should all of the files in it be
+        included from the scan?
+
+        :param dir_path: The path of the directory, relative to the root directory
+                         being scanned.  This is a directory, so the path will
+                         always end in '/'.
+        :return: True iff excluded.
+        """
+
+    def should_include(self, path):
+        if path.endswith('/'):
+            return self.should_include_directory(path)
+        return self.should_include_file(path)
+
+
+class ScanExcludeDirRegex(ScanExcludePolicy):
+    def __init__(self, regex):
+        self.regex_str = regex
+        self.regex = re.compile(regex)
+
+    def __repr__(self):
+        return 'ExcludeDirRegex(%s)' % (self.regex_str,)
+
+    def should_exclude_file(self, file_path):
+        head, _ = os.path.split(file_path)
```

Comment:
The names we're dealing with here are B2 names, which always use `/` as the separator.  `os.path.split` will want to use `\` if you're on Windows.